src/MongoDB.Driver.Core/Core/Operations/RetryableWriteOperationExecutor.cs

```diff
@@ -145,6 +145,7 @@ private static bool DoesContextAllowRetries(RetryableWriteContext context)
             return
                 context.RetryRequested &&
                 AreRetryableWritesSupported(context.Channel.ConnectionDescription) &&
+                context.Binding.Session.Id != null &&
```

Comment:
The session Id will only be present if ALL mongos instances report a value for LogicalSessionTimeoutMinutes.

Or you can think of this in terms of the server error: a transaction number (i.e. a retry) is only allowed when a sessionId is also specified.

Reply:
Any way this could be unit tested?

Reply:
Yes. I will write some unit tests for these helper methods.

It's only an integration test for the full scenario with real servers that is hard/impossible to write.

Reply:
There was an existing unit test for this method so I just refactored it to test that the session Id is also taken into account.

Reply:
I would suggest adding a comment for this line since the reference between this change and `LogicalSessionTimeoutMinutes` is not obvious

Reply:
We can ignore the relationship to LogicalSessionTimeout. The only that matters is whether `Session.Id` has a value or not.

I have edited the JIRA ticket to phrase the issue in terms of session Id also.